src/number_theory/legendre_symbol/mul_character.lean

```diff
@@ -0,0 +1,495 @@
+def is_quadratic (χ : mul_char R R') : Prop := ∀ a, χ a = 0 ∨ χ a = 1 ∨ χ a = -1
+
+/-- We can post-compose a multiplicative character with a ring homomorphism. -/
+def ring_hom_comp (χ : mul_char R R') (f : R' →+* R'') : mul_char R R'' :=
+{ map_nonunit' :=
+  begin
+    intros a ha,
+    simp only [ring_hom.to_monoid_hom_eq_coe, monoid_hom.to_fun_eq_coe, monoid_hom.coe_comp,
+               ring_hom.coe_monoid_hom, coe_coe, function.comp_app, map_nonunit χ ha,
+               ring_hom.map_zero],
+  end,
+  ..f.to_monoid_hom.comp χ.to_monoid_hom}
+
+@[simp]
+lemma ring_hom_comp_apply (χ : mul_char R R') (f : R' →+* R'') (a : R) :
+  χ.ring_hom_comp f a = f (χ a) := rfl
+
+/-- Composition with an injective ring homomorphism preserves nontriviality. -/
+lemma is_nontrivial.comp {χ : mul_char R R'} (hχ : χ.is_nontrivial)
+  {f : R' →+* R''} (hf : function.injective f) :
+   (χ.ring_hom_comp f).is_nontrivial :=
+begin
+  obtain ⟨a, ha⟩ := hχ,
+  use a,
+  rw [ring_hom_comp_apply, ← ring_hom.map_one f],
+  exact λ h, ha (hf h),
+end
+
+/-- Composition with a ring homomorphism preserves the property of being a quadratic character. -/
+lemma is_quadratic.comp {χ : mul_char R R'} (hχ : χ.is_quadratic) (f : R' →+* R'') :
+  (χ.ring_hom_comp f).is_quadratic :=
+begin
+  intro a,
+  simp only [ring_hom_comp_apply],
+  rcases hχ a with (ha | ha | ha); rw [ha],
+  { left,
+    rw [ring_hom.map_zero], },
+  { right, left,
+    rw [ring_hom.map_one], },
+  { right, right,
+    rw [ring_hom.map_neg, ring_hom.map_one], },
+end
+
+/-- The inverse of a quadratic character is itself. →  -/
+lemma quadratic_char_inv {χ : mul_char R R'} (hχ : χ.is_quadratic) : χ⁻¹ = χ :=
+begin
+  ext x,
+  rw [inv_apply_eq_inv],
+  rcases hχ x with h₀ | h₁ | h₂,
+  { rw [h₀, ring.inverse_zero], },
+  { rw [h₁, ring.inverse_one], },
+  { rw [h₂, (by norm_num : (-1 : R') = (-1 : R'ˣ)), ring.inverse_unit (-1 : R'ˣ)],
+    refl, },
+end
+
+/-- The square of a quadratic character is the trivial character. -/
+lemma quad_char_sq_eq_one {χ : mul_char R R'} (hχ : χ.is_quadratic) : χ ^ 2 = 1 :=
+begin
+  rw [pow_two],
+  nth_rewrite 0 ← quadratic_char_inv hχ,
+  rw [mul_left_inv],
+end
+
+/-- The `p`th power of a quadratic character is itself, when `p` is the (prime) characteristic
+of the target ring. -/
+lemma quad_char_pow_char {χ : mul_char R R'} (hχ : χ.is_quadratic)
+ (p : ℕ) [hp : fact p.prime] [char_p R' p] :
+  χ ^ p = χ :=
+begin
+  ext x,
+  rw [pow_apply_coe],
+  rcases hχ x with (hx | hx | hx); rw hx,
+  { rw [zero_pow (fact.out p.prime).pos], },
+  { rw [one_pow], },
+  { exact char_p.neg_one_pow_char R' p, },
+end
+
+/-- The `n`th power of a quadratic character is itself, when `n` is odd. -/
+lemma quad_char_pow_odd {χ : mul_char R R'} (hχ : χ.is_quadratic) {n : ℕ} (hn : n % 2 = 1) :
+  χ ^ n = χ :=
+by rw [← nat.div_add_mod n 2, hn,
+       pow_add, pow_mul, quad_char_sq_eq_one hχ, one_pow, pow_one, mul_char.one_mul]
```

Comment:
I'd split this in the even and odd part, since we're basically proving the even part already:
```suggestion
/-- The `n`th power of a quadratic character is the trivial character, when `n` is even. -/
lemma is_quadratic.pow_even {χ : mul_char R R'} (hχ : χ.is_quadratic) {n : ℕ} (hn : even n) :
  χ ^ n = 1 :=
begin
  obtain ⟨n, rfl⟩ := even_iff_two_dvd.mp hn,
  rw [pow_mul, hχ.sq_eq_one, one_pow]
end

/-- The `n`th power of a quadratic character is itself, when `n` is odd. -/
lemma quad_char_pow_odd {χ : mul_char R R'} (hχ : χ.is_quadratic) {n : ℕ} (hn : odd n) :
  χ ^ n = χ :=
begin
  obtain ⟨n, rfl⟩ := hn,
  rw [pow_add, pow_one, hχ.pow_even (even_two_mul _), one_mul]
end
```

Reply:
I guess you want `is_quadratic.pow_odd` here, same as for even etc.